### PR TITLE
Attestation is Decoupled From Proposal

### LINKED ISSUE #3796: One validator in a committee usually attests to a different data root
Example logs from testnet:
```
[2019-10-17 09:52:42] DEBUG operation: block attestation aggregation_bits=[       0      100] index=0 root=0x5b0396ccd538519b93d814436992caf7871ab05f3cbbea4a7005789661d451da shard=3                              
[2019-10-17 09:52:42] DEBUG operation: block attestation aggregation_bits=[       0      100] index=1 root=0x6eef55ae1cc1b97ff5e7aec332c9411fa7c29172119523c4ebddeb6532cfa491 shard=4                              
[2019-10-17 09:52:42] DEBUG operation: block attestation aggregation_bits=[11111111 11111011] index=2 root=0xe1774dcbdc0dcc2451b1c05de1d1ba9b5a76a82e3fc0fa1c8bb92f2c4ab0911b shard=3                              
[2019-10-17 09:52:42] DEBUG operation: block attestation aggregation_bits=[11111111 10011011] index=3 root=0x34c25742503f2d18f0acd510677a4c60bcf0151e9bdce2bef07adc204aecbf11 shard=4 
```
The expected behavior is that the committee (identified by shard in this example) should have all validators attesting to the same root. 

This is consistently reproducible in a single client local network with 64 validators. Turn debug logs on to see information about the block attestations printed when the operations service prunes attestations from the pool.

### LINKED ISSUE #3801: Attester attests to head at last slot
We are seeing attester attests to the head for last slot not current slot.
What we want:
```
---------------slot N-------------
| propose head N | attest head N |
```
What we see:
```
----------------slot N----------------
| propose head N | attest head N - 1 |
```

### PR DESCRIPTION
**Pull request by nisdas**
This resolves #3796 and resolves #3801 . The main bug was that we did not wait till the slot midpoint before attesting for the current head. Instead all attestations and block proposals happened in the same timeframe.
- [x] Decouples attestation and proposal duties.
- [x] Passes in the time as second instead of ns.
- [x] Adds a regression test

### CODE REVIEW

File: slasher/rpc/server_test.go [hunk 1]
```diff
@@ -2,9 +2,10 @@ package rpc
 
 import (
 	"context"
-	"github.com/gogo/protobuf/proto"
 	"testing"
 
+	"github.com/gogo/protobuf/proto"
```
Comment (terencechain) [inline slasher/rpc/server_test.go]: imports

**Comment by prestonvanloon** [comment]
Please add details to the PR description. What was the problem?

File: validator/client/validator_attest.go [hunk 2]
```diff
@@ -135,7 +135,7 @@ func (v *validator) waitToSlotMidpoint(ctx context.Context, slot uint64) {
 	if half == 0 {
 		delay = 500 * time.Millisecond
 	}
-	duration := time.Duration(slot*params.BeaconConfig().SecondsPerSlot) + delay
+	duration := time.Second*time.Duration(slot*params.BeaconConfig().SecondsPerSlot) + delay
```
Comment (prestonvanloon) [inline validator/client/validator_attest.go]: 😆

**Comment by prestonvanloon** [comment]
Please resolve conflicts

File: validator/client/validator_attest.go [hunk 3]
```diff
@@ -135,7 +135,13 @@ func (v *validator) waitToSlotMidpoint(ctx context.Context, slot uint64) {
...
 		delay = 500 * time.Millisecond
 	}
-	duration := time.Duration(slot*params.BeaconConfig().SecondsPerSlot) + delay
-	timeToBroadcast := time.Unix(int64(v.genesisTime), 0).Add(duration)
+	startTime := slotStartTime(v.genesisTime, slot)
+	timeToBroadcast := startTime.Add(delay)
 	time.Sleep(roughtime.Until(timeToBroadcast))
 }
+
+func slotStartTime(genesis uint64, slot uint64) time.Time {
```
Comment (terencechain) [inline validator/client/validator_attest.go]: would this be better under `/helpers/slot.go`?

Comment (nisdas) [reply validator/client/validator_attest.go]: good idea, I will shift it there

File: validator/client/runner.go [hunk 4]
```diff
@@ -92,8 +92,8 @@ func run(ctx context.Context, v Validator) {
 					case pb.ValidatorRole_ATTESTER:
 						v.AttestToBlockHead(slotCtx, slot, id)
 					case pb.ValidatorRole_PROPOSER:
+						go v.AttestToBlockHead(slotCtx, slot, id)
```
Comment (prestonvanloon) [inline validator/client/runner.go]: nice

File: validator/client/validator_attest.go [hunk 5]
```diff
@@ -135,7 +136,7 @@ func (v *validator) waitToSlotMidpoint(ctx context.Context, slot uint64) {
 	if half == 0 {
 		delay = 500 * time.Millisecond
 	}
-	duration := time.Duration(slot*params.BeaconConfig().SecondsPerSlot) + delay
-	timeToBroadcast := time.Unix(int64(v.genesisTime), 0).Add(duration)
+	startTime := slotutil.SlotStartTime(v.genesisTime, slot)
+	timeToBroadcast := startTime.Add(delay)
```
Comment (prestonvanloon) [inline validator/client/validator_attest.go]: much better, thank you

**Pull request merged**